### PR TITLE
Introduce IntervalPartition algorithm

### PR DESCRIPTION
**Pull request by tsreaper**

### CODE REVIEW

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import org.apache.flink.table.storage.filestore.lsm.SortedRun;
+import org.apache.flink.table.storage.filestore.lsm.sst.SstFileMeta;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+/** Algorithm to partition several sst files into the minimum number of {@link SortedRun}s. */
+public class IntervalPartition {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java:32]: Why not the input is sst files?
So there is another class to produce `SortedRun`?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                (o1, o2) -> {
+                    int leftResult = keyComparator.compare(o1.minKey(), o2.minKey());
+                    return leftResult == 0
+                            ? keyComparator.compare(o1.maxKey(), o2.maxKey())
+                            : leftResult;
+                });
+        this.keyComparator = keyComparator;
+    }
+
+    public List<List<SortedRun>> partition() {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java:66]: What is the output? Can you add document for this?
What is difference between `partition()` and `partition(metas)`?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/sst/SstFileMeta.java [hunk 3]
```diff
@@ -130,7 +130,9 @@ public int hashCode() {
                 rowCount,
                 minKey,
                 maxKey,
-                stats,
+                // by default, hash code of arrays are computed by reference, not by content.
+                // so we must use Arrays.hashCode to hash by content.
+                Arrays.hashCode(stats),
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/sst/SstFileMeta.java:135]: Can you add test for this?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java (hunk 1 above)
Comment (tsreaper) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java:32]: `SortedRun` ensures that the files are sorted. We can, of course, accept a list of files and sort them by ourselves but in an lsm for all levels >= 1 the files are already sorted and I think we can take advantage of this.

Let's change the input for now and come back to this if we need this optimization.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java (hunk 2 above)
Comment (tsreaper) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java:66]: `partition()` is the public interface. It returns several sections (outer list) of sorted runs (inner list). `partition(metas)` is the inner implementation for a specific section. I'll add some documentation.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/sst/SstFileMeta.java (hunk 3 above)
Comment (tsreaper) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/sst/SstFileMeta.java:135]: This is already used by `IntervalPartitionTest`. Without this, the comparison will fail.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+/** Algorithm to partition several sst files into the minimum number of {@link SortedRun}s. */
+public class IntervalPartition {
+
+    List<SstFileMeta> files;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java]: private final

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    RecordComparator keyComparator;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java]: ditto

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    public IntervalPartition(List<SstFileMeta> sortedRuns, RecordComparator keyComparator) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/IntervalPartition.java]: sortedRuns -> files?
If we want to compact two levels, we'd better break up their files and repartition them according to the IntervalPartition to come up with the optimal strategy.

**Review by JingsongLi — APPROVED** [review]
Left final comments

**Pull request merged**